erts/emulator/beam/jit/arm/instr_arith.cpp

```diff
@@ -1519,11 +1519,18 @@ void BeamModuleAssembler::emit_i_bsr(const ArgLabel &Fail,
 
         /* Ensure that both operands are small and that the shift
          * count is positive. */
+        ERTS_CT_ASSERT(_TAG_IMMED1_SMALL == _TAG_IMMED1_MASK);
         a.and_(TMP1, lhs.reg, rhs.reg);
-        a.mvn(TMP1, TMP1);
-        a.tst(TMP1, imm(_TAG_IMMED1_MASK));
-        a.ccmp(rhs.reg, imm(0), imm(NZCV::kSigned), arm::CondCode::kEQ);
-        a.b_lt(generic);
+        if (std::get<0>(getClampedRange(RHS)) >= 0) {
+            a.and_(TMP1, TMP1, imm(_TAG_IMMED1_MASK));
+            a.cmp(TMP1, imm(_TAG_IMMED1_SMALL));
+            a.b_ne(generic);
+        } else {
+            a.mvn(TMP1, TMP1);
+            a.tst(TMP1, imm(_TAG_IMMED1_MASK));
+            a.ccmp(rhs.reg, imm(0), imm(NZCV::kSigned), arm::CondCode::kEQ);
+            a.b_lt(generic);
```

Comment:
Slightly shorter, assuming we move the initial `a.and_(TMP1, lhs.reg, rhs.reg);` into the other branch:
```suggestion
            a.ands(TMP1, rhs.reg, imm((1ul << 63) | _TAG_IMMED1_SMALL));
            a.and_(TMP1, lhs.reg, TMP1);
            a.ccmp(TMP1, imm(_TAG_IMMED1_SMALL), imm(NZCV::kNone), arm::CondCode::kPL);
            a.b_ne(generic);
```

Reply:
Nice! I removed the other branch since it is no longer any shorter.